app/uk/gov/hmrc/plasticpackagingtax/registration/models/registration/OrganisationDetails.scala

```diff
@@ -33,7 +43,31 @@ case class OrganisationDetails(
   soleTraderDetails: Option[SoleTraderIncorporationDetails] = None,
   incorporationDetails: Option[IncorporationDetails] = None,
   partnershipDetails: Option[PartnershipDetails] = None
-)
+) {
+
+  def status: TaskStatus =
+    if (isBasedInUk.isEmpty) TaskStatus.NotStarted
+    else if (businessPartnerIdPresent()) TaskStatus.Completed
+    else TaskStatus.InProgress
+
+  def businessPartnerIdPresent() =
+    organisationType match {
+      case Some(UK_COMPANY) if incorporationDetails.isDefined =>
+        incorporationDetails.get.registration.registeredBusinessPartnerId.isDefined
```

Comment:
Instead of guard on case and then using "get" (which always smells IMO) you could do something like this I think
```
      case Some(UK_COMPANY) =>
        incorporationDetails.map(_.registration.registeredBusinessPartnerId.isDefined)
```

Reply:
I don't think that will work here, not in its current form at least, as I want to return a Boolean from this function. I suspect there is a cleaner, more Scala-friendly alternative to what I have but I can't see it off the top of my head. Any further suggestions?

Reply:
You could use "exists".  E.g.
```
  def businessPartnerIdPresent() =
    organisationType match {
      case Some(UK_COMPANY) =>
        incorporationDetails.exists(_.registration.registeredBusinessPartnerId.isDefined)
      case Some(SOLE_TRADER) =>
        soleTraderDetails.exists(_.registration.registeredBusinessPartnerId.isDefined)
      case Some(PARTNERSHIP) if partnershipDetails.isDefined =>
        partnershipDetails.exists(details => details match {
          case GENERAL_PARTNERSHIP =>
            details.generalPartnershipDetails.exists(_.registration.registeredBusinessPartnerId.isDefined)
          case SCOTTISH_PARTNERSHIP =>
            details.scottishPartnershipDetails.exists(_.registration.registeredBusinessPartnerId.isDefined)
          case _ => false
        })
      case _ => false
    }
```

Reply:
I think that reads marginally better. I'll go with that.